Refactor async completion logic for read and write path [run-systemtest]

bjorncs opened the pull request:
Ensure that failure from request read, write or async context correctly terminates request processing.

FYI @gjoranv @jonmv

bjorncs committed:
Refactor async completion logic for read and write path

Ensure that failure from request read, write or async context correctly terminates request processing.

baldersheim reviewed (approved):
Fix close, and then factory will do its deed.

bjorncs committed:
Ensure writer is closed while holding lock

Handler can inject its response content if lock is not held between write and close.
